### ISSUE TITLE
investigate GPU conservation

### ISSUE THREAD

**Issue by juliasloan25** [CliMA/ClimaCoupler.jl#607]
Our GPU runs have higher water and energy conservation than CPU runs with identical setups. The CPU error tends to be around `1e-5`, while the GPU error is around `1e-3` or `1e-4`. We should look into why the conservation is worse on GPU, and try to improve it.

For now (as of #589), we've set the GPU runs to soft fail if the conservation error is larger than `1e-3`, but we want them to be able to pass this threshold (and even a smaller one, ideally).

part of #390 
PR #614

Partially done in https://github.com/CliMA/ClimaCoupler.jl/issues/735

### Approaches to try
- add debug plots, as was done in [this branch](https://github.com/CliMA/ClimaCoupler.jl/compare/main...ln/interactive-debug)
  - requires ability to convert fields from GPU to CPU - see https://github.com/CliMA/ClimaCore.jl/pull/1619
- check float type stability
  - Atmos cache always initialized with NaN
  - Land state always initialized with Float64 in coupler
  - might not be relevant if our runs are already with Float64, but could affect Float32 runs

### Bucket info
ClimaLand has standalone global bucket runs on CPU and GPU that we compare. For all 3 albedo options, comparing the mean values of the states from CPU and GPU runs gives a difference on the order of 1e-15. For the temporal map albedo case, which runs for 50 days, we get:
```
mean(cpu_state .- gpu_state) = 1.433428363408908e-15
eps(Float64) = 2.220446049250313e-16
```
The functional and static map albedo cases show similar discrepancies, and run for 7 days each. These differences are much smaller than what we see in coupled runs, so the difference is probably not coming from the bucket model.

#### functional albedo
CPU
rse[end] = 1.6532404532423364e-5
rse[end] = 0.0005462468417393023
GPU
rse[end] = 7.841693994284924e-5
rse[end] = 0.0005103673696554503

 <img src="https://github.com/CliMA/ClimaCoupler.jl/assets/51397186/cfcfd42a-8dc4-459f-82ef-507a8d453936" alt="CPU analytic function" width="800"/>

#### static map
CPU
rse[end] = 1.5598104032642762e-5
rse[end] = 7.682827509990757e-5
GPU
rse[end] = 0.00012549802174549964
rse[end] = 0.0015490000882394841

 <img src="https://github.com/CliMA/ClimaCoupler.jl/assets/51397186/356e2fe4-d73e-4b51-9544-ef08657fe54c" alt="CPU static map" width="400"/>
 <img src="https://github.com/CliMA/ClimaCoupler.jl/assets/51397186/db2daa42-44ce-4330-9de4-4c86680176bd" alt="GPU static map" width="400"/>

#### temporal map
CPU
rse[end] = 1.7424794137292602e-5
rse[end] = 0.00033606441246576436
GPU
rse[end] = 0.00012708219201050294
rse[end] = 0.000943450817881417

 <img src="https://github.com/CliMA/ClimaCoupler.jl/assets/51397186/81c717b3-8124-40f5-9e2f-0549d86af158" alt="CPU temporal map" width="400"/>
 <img src="https://github.com/CliMA/ClimaCoupler.jl/assets/51397186/450fe3dd-9fd0-4dd0-9563-c0ffe4ba4391" alt="GPU temporal map" width="400"/>

**Comment by akshaysridhar** [comment]
These might be relevant builds (ClimaAtmos - standalone test cases) 
https://buildkite.com/clima/climaatmos-ci/builds/16728#018da910-054b-4a34-aed1-0807b98db143 (HS)
https://buildkite.com/clima/climaatmos-ci/builds/16728#018da910-054b-4a34-aed1-0807b98db143 (BW)

**Comment by LenkaNovak** [comment]
See stand-alone Moist held-suarez atmos runs here:
https://buildkite.com/clima/climacoupler-longruns/builds/514

Some plots:

Zonal mean wind
![Screen Shot 2024-04-04 at 7 07 43 PM](https://github.com/CliMA/ClimaCoupler.jl/assets/38083334/20ee0450-1288-4473-808f-d2c69793d6d2)

Zonal mean temperature
![Screen Shot 2024-04-04 at 7 08 11 PM](https://github.com/CliMA/ClimaCoupler.jl/assets/38083334/ed488e6e-28c6-43ee-9495-5d4d967e52c1)

1st level temperature
![Screen Shot 2024-04-04 at 7 09 00 PM](https://github.com/CliMA/ClimaCoupler.jl/assets/38083334/d997ec06-a567-42d6-9ed4-1e9dbded4e94)

**Comment by juliasloan25** [comment]
`Fields.bycolumn` is effectively a no-op on GPU and could be contributing to the differences between CPU and GPU. See https://github.com/CliMA/ClimaCoupler.jl/issues/736

**Comment by LenkaNovak** [comment]
Thanks to #737, we can confirm that our GPU longruns run as well as CPU longruns ([build](https://buildkite.com/clima/climacoupler-longruns/builds/619#_)). Conservation logging issue (20% error) was addressed in #735. The GPU runs are systematically less conservative than the CPU runs (discussion in #735) but the difference is quite small (~1%), so we can revisit this in the future (as part of #594) when we can track energy sinks and sources to the precision of sqrt(eps) - this requires work in ClimaAtmos (see https://github.com/CliMA/ClimaAtmos.jl/issues/2658 and https://github.com/CliMA/ClimaAtmos.jl/issues/2568), and close down this issue. @juliasloan25, would you agree?

**Comment by LenkaNovak** [comment]
completed
